src/taskorchestrationexecutor.ts

```diff
@@ -0,0 +1,493 @@
+            error = new OrchestrationFailureError(this.orchestratorReturned, orchestratorState);
+            result = undefined;
+        }
+
+        // Communicate the orchestration's current state
+        context.done(error, result);
+        return;
+    }
+
+    /**
+     * @hidden
+     * Determine if the orchestrator should exit, either successfully or through an error.
+     *
+     * @returns
+     *  True if the orchestration's invocation completed, or if an unhandled exception was thrown.
+     *  False otherwise.
+     */
+    private hasExecutionCompleted(): boolean {
+        return this.orchestrationInvocationCompleted() || this.exception !== undefined;
+    }
+
+    /**
+     * @hidden
+     * Determine if the current invocation has finished.
+     *
+     * @returns
+     *  True if the orchestration reached a `return` statement, or a `continueAsNew`.
+     *  False otherwise.
+     */
+    private orchestrationInvocationCompleted(): boolean {
+        return this.orchestratorReturned || this.willContinueAsNew;
+    }
+
+    /**
+     * @hidden
+     * Processes a History event, often by either by updating some deterministic API value, updating
+     * the state of a task, or resuming the user code.
+     *
+     * @param event
+     *  The History event we're currently processing
+     */
+    private processEvent(event: HistoryEvent): void {
+        const eventType = event.EventType;
+        switch (event.EventType) {
+            case HistoryEventType.OrchestratorStarted: {
+                const timestamp = event.Timestamp;
+                if (timestamp > this.context.currentUtcDateTime) {
+                    this.context.currentUtcDateTime = timestamp;
+                }
+                break;
+            }
+            case HistoryEventType.ContinueAsNew: {
+                // The clear all state from the orchestration,
+                // as if no processing of History had taken place
+                this.initialize();
+                break;
+            }
+            case HistoryEventType.ExecutionStarted: {
+                this.tryResumingUserCode();
+                break;
+            }
+            case HistoryEventType.EventSent: {
+                // The EventSent event requires careful handling because it is re-used among
+                // CallEntity and WaitForExternalEvent APIs.
+                // For CallEntity, the EventRaised event that contains that API's result will
+                // expect a TaskID that is different from the TaskID found at the root of this
+                // EventSent event. Namely, the TaskID it expects can be found nested in the
+                // "Input" field of the corresponding EventSent event. Here, we handle that
+                // edge-case by correcting the expected TaskID in our openTask list.
+                const key = event.EventId;
+                if (key in this.openTasks) {
+                    const task = this.openTasks[key];
```

Comment:
Are the above two lines redundant? Could we save some cycles by fetching the task and checking for `undefined`?  I noticed this in a few other places as well.

Reply:
You mean to eagerly index into the array, and then check for undefined? Yeah we could do that, but I'm not particularly fond of how that coding style. Also, in my experience, it takes a lot of convincing to the TypeScript type-checker that a value is no longer `undefined`, so it can make the code harder to read.

Perhaps we can find a happy middle-ground by utilizing some custom utility. Let me think about this.

Reply:
Fair enough, I'll make the change in all appropiate places (there's about 4 in this file) :) . My next commit includes them.

Tangentially, I wish TypeScript / JS had better support for this scenario - I find it really weird that there's no idiomatic way to safely index into an object without risking nullish types snealing into the code. But oh well, that's a concern for the TypeScript team, not for this PR.

I'm changing this for the performance-friendly alternative